src/worker.rs

```diff
@@ -264,21 +293,39 @@ impl Worker {
                 self.kill_handle.abort();
                 (
                     self.spec,
-                    Some(Arc::new(anyhow::Error::new(termination_timeout_err))),
+                    Some(Arc::new(TerminationError::TimeoutError(
+                        termination_timeout_err,
+                    ))),
                 )
             }
 
-            // abort logic failed for some reason
-            Ok(Err(abort_err)) => (self.spec, Some(Arc::new(anyhow::Error::new(abort_err)))),
-
-            // worker routine returned with a failure, but because this is
-            // a termination, we return the error
-            Ok(Ok(Err(routine_err))) => {
-                (self.spec, Some(Arc::new(anyhow::Error::new(routine_err))))
+            //
+            Ok(Err(join_handle_err)) => {
+                if join_handle_err.is_panic() {
+                    (
+                        self.spec,
+                        Some(Arc::new(TerminationError::PanicError(join_handle_err))),
+                    )
+                } else {
+                    (
+                        self.spec,
+                        Some(Arc::new(TerminationError::CancelError(join_handle_err))),
+                    )
+                }
             }
 
+            Ok(Ok(Err(kill_err))) => (
+                self.spec,
+                Some(Arc::new(TerminationError::KillError(kill_err))),
+            ),
+
+            Ok(Ok(Ok(Err(termination_err)))) => (
+                self.spec,
+                Some(Arc::new(TerminationError::ClientError(termination_err))),
+            ),
+
             // happy path
-            Ok(Ok(Ok(_))) => return (self.spec, None),
+            Ok(Ok(Ok(Ok(_)))) => (self.spec, None),
```

Comment:
> Are you familiar with the Try trait and the ? operator for bubbling up errors? 

Barely, I went ahead and investigated a bit more. It is pretty convenient (and somewhat magical) how the implicit cast of errors work. Not sure if I like it, but it seems to make things better (wait and see to form a solid opinion). Thanks for pointing me to that

> I wonder if there is a way to start unifying the error types earlier and unwrap some of these layers of nesting.

I started to go through that, and it is not as easy given how this nesting gets built. It is part of the usage of each future combinator call, unless I'm missing some nifty function that allows to do a `(->)` with error composition, I don't think we can get away with it (Would love to be wrong though).

> I think it makes sense to not introduce the Arc at this level since it is not actually being shared yet, and it will make it a lot cleaner to deal with these errors. 

Agreed, I'll get rid of it

Reply:
The [`TryFutureExt`](https://docs.rs/futures/0.3.5/futures/future/trait.TryFutureExt.html) trait might be helpful for composing future combinators in a way that reduces the nesting. However, that won't help with the timeout.  And all of these layers of results do represent different errors.  Definitely not required to fix in this PR though I think.

Reply:
Cool, let me know if it is ok to merge, want to get started with the supervision system events, I'll use that to test supervision logic in a black-box fashion